### ISSUE TITLE
[enhancement request] More levels of detail in Coq's error messages

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#6201]
#### Version

8.7.0

#### Description of the problem

Frequently it seems that error messages need to target multiple audiences at once.  Notorious examples are failures of `pattern` due to dependent types, `setoid_rewrite` error messages, and unification failures.  

Error messages like "rewrite failed: could not guess a way to make the dependent types line up" seems reasonable for a beginner, who would be confused by a wall of text.  (I, on the other hand, very much value the "could not be applied to arguments ... because the nth argument has type foo while it's expected to have type bar" error message, because it tells me exactly what I need to know, and the context is there for me to dig into if I need it.)

Similarly, "setoid_rewrite failed: Did you declare the relation eq as compatible with iff in its first argument via a Proper typeclass instance?" seems like a very good error message for a beginner (c.f. #6141), but is utterly useless to me (as is the current error message); I want to know where typeclass resolution failed, and how, possibly with a bit of detail around the first failure, but definitely not the full typeclass log (see also https://github.com/asya-bergal/setoid_rewrite_debug).

TeX has multiple levels for most error messages, e.g., you might get
> Infinite glue shrinkage found on current page

and if you ask for more detail, TeX says
> The page about to be output contains some infinitely
> shrinkable glue, e.g., `\vss` or `\vskip 0pt minus 1fil`.
> Such glue doesn't belong there; but you can safely proceed,
> since the offensive shrinkability has been made finite.

It might be nice to be able to get such variability in error messages I get from Coq, e.g., maybe `Set Error Messages Detailed` vs `Set Simple Error Messages` or something.

**Comment by ejgallego** [comment]
We are close to the point we could try to implement something like that, even if I would like to get the STM error handling in order.

I guess we could imitate the warning and define a type of error message and register it with the proper code. The error should then take all the possibly needed arguments.

I am a fan of declaring errors as an inductive, but that seems not so popular in Coq.

**Comment by herbelin** [comment]
> I am a fan of declaring errors as an inductive, but that seems not so popular
in Coq.

What does "declaring errors as an inductive" mean?

**Comment by ejgallego** [comment]
> What does "declaring errors as an inductive" mean?

I mean having all errors go throu a type:
```
type toplevel_error =
| Missing_Lib of ...
|
```
plus the corresponding "explain" function. This is done in some parts, but we could maybe go further.

**Comment by ppedrot** [comment]
@ejgallego I agree for well-delimited components, but beware that this makes a fundamentally closed-world assumption (universal properties of inductive types, blah blah). So you can't really have one type describing all errors for Coq. That said, note that it's almost already the case component-wise, we have the pretype errors, typing errors, etc.

**Comment by ejgallego** [comment]
Yes you can, just add a catch-all constructor that would play the role of UserError today.

[You can also do fancier stuff with extensible types, but that's not the point.

**Comment by ppedrot** [comment]
> Yes you can, just add a catch-all constructor that would play the role of UserError today.

But then I don't understand the advantage of this approach. We also want people raising `UserError` to have various levels of explanations, don't we?

**Comment by ejgallego** [comment]
Today you cannot raise other than `UserError of Pp.t`, so it is hard to:
- give printing explanations.
- be lazier on printing.
- provide an error index for the manual that can be automatically processed.

what I propose enrich the things that can be raised, with a catch-all constructor for today's use.

I don't know how to implement several levels of explanations when catching an `UserError`.

**Comment by ppedrot** [comment]
> Today you cannot raise other than UserError of Pp.t

Not quite. I routinely register toplevel error handlers in `CErrors` in order to perform lazy printing.

**Comment by ejgallego** [comment]
> Not quite. I routinely register toplevel error handlers in CErrors in order to perform lazy printing.

Well, the introduction of that mechanism was not followed with the migration of the normal error printing to it, in fact I think we are closer of removing printer registration than from going in the direction of migrating everything.

There is more discussion about the status of `iprint` in ·5524 #5360

**Comment by Zimmi48** [comment]
You can also have a generic handling of several levels of details, e.g. with a list of thunks...

BTW I'm very happy with the general will to support better error messages (and the many reports and PRs already in [`kind: user messages`](https://github.com/coq/coq/issues?utf8=%E2%9C%93&q=label%3A%22kind%3A%20user%20messages%22%20)). When it comes to inspiration on how to make better error messages, I often turn to Elm: http://elm-lang.org/blog/compilers-as-assistants
